### PR TITLE
[libc++] Document guidelines for applying [[nodiscard]]

### PR DESCRIPTION
**Pull request by philnik777**
We've been applying ``[[nodiscard]]`` more liberally recently, but we don't have any documented guidance on when it's correct to add it. This patch adds that guidance. Follow-up patches will gradually apply it to the code base.

### CODE REVIEW

**Comment by philnik777** [comment]
CC @ldionne @mordante @cjdb I'm sure you have some thoughts on this.

**Comment by mordante** [comment]
First of all thanks for working on this!
As mentioned in #83695 there is a paper [P3122R0](https://wg21.link/p3122r0) "[[nodiscard]] should be Recommended Practice" by @jwakely and there is another paper [P2422R0](https://wg21.link/p2422r0) "Remove nodiscard annotations from the standard library specification" that proposes to drop all `[[nodiscard]]`s from the draft.

These papers are likely to be discussed in a few weeks during the Tokyo meeting. So I would propose to wait until that has been discussed; it might affect this RFC. Even when these papers are not accepted we should look at P3122R0 it has recommended practices we could use as guideline in our documentation.

Note I'm in favour of using `[[nodiscard]]` more in libc++ and the direction of these papers.

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+- which only observe a value. For example ``string::size()``.
+
+  Code that discards values from these kinds of functions is dead code. It can
+  either be removed, or the programmer meant to do something different.
+
+- where discarding the value is most likely a misuse of the function. For
+  example ``find``.
+
+  This protects programmers from assuming too much about how the internals of
+  a function work, resulting in less code breaking as optimizations are applied.
```
Comment (hawkinsw) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:34]: Thank you for adding this documentation!

```suggestion
  a function work, making code more robust in the presence of future optimizations.
```

**Comment by philnik777** [comment]
> First of all thanks for working on this! As mentioned in #83695 there is a paper [P3122R0](https://wg21.link/p3122r0) "[[nodiscard]] should be Recommended Practice" by @jwakely and there is another paper [P2422R0](https://wg21.link/p2422r0) "Remove nodiscard annotations from the standard library specification" that proposes to drop all `[[nodiscard]]`s from the draft.
> 
> These papers are likely to be discussed in a few weeks during the Tokyo meeting. So I would propose to wait until that has been discussed; it might affect this RFC. Even when these papers are not accepted we should look at P3122R0 it has recommended practices we could use as guideline in our documentation.
> 
> Note I'm in favour of using `[[nodiscard]]` more in libc++ and the direction of these papers.

Thanks for bringing them up. I agree that this may change things, but I also think that discussing this beforehand within the library also makes a lot of sense, since "look, one implementation already does it, the second one is working on getting `[[nodiscard]]` in lots of places, and the third one's main maintainer is proposing this stuff" is a really good argument (got a bit wordy). If we talk about it beforehand, we have thought about it already quite a bit and can hopefully give good reasons to add `[[nodiscard]]` to some functions and not to others.

From looking at the list of P3122R0 I think we want to add comparison operators. Everything else should already be covered™.

FWIW I also very much agree with the sentiment of the papers.

**Comment by jwakely** [comment]
The whole point of my [P3122R0](https://wg21.link/p3122r0) paper is that I think implementations should make their own decisions about where to use nodiscard. The standard should do no more than make some vague recommendations, and definitely *not* try to standardize where it should be used (or even worse, standardize where it should not be used).

So I am very pleased to see this change and the start of a discussion. Please don't wait for WG21 to specify nodiscard on individual functions or on some minimal "low effort" set of functions.

**Comment by mordante** [comment]
> The whole point of my [P3122R0](https://wg21.link/p3122r0) paper is that I think implementations should make their own decisions about where to use nodiscard. The standard should do no more than make some vague recommendations, and definitely _not_ try to standardize where it should be used (or even worse, standardize where it should not be used).

True, but your paper has some nice suggestions and when that becomes recommended practice I would like to see that as minimal set in our documentation. After which we can look whether more places would be useful to have nodiscards. 

> So I am very pleased to see this change and the start of a discussion. Please don't wait for WG21 to specify nodiscard on individual functions or on some minimal "low effort" set of functions.

At the moment all our nodiscards not specified in the Standard can be disabled. The disabling mechanism  tests all occurrences. Depending on the outcome in discussion Tokyo I want to discuss internally whether we want to keep testing all disabling occurrences or not. (We already removed the documentation for them.)

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+======================================
+
+Libc++ adds ``[[nodiscard]]`` to functions in a lot more places than the
+standard does. Any applications of ``[[nodiscard]]`` that aren't required by the
+standard written as ``_LIBCPP_NODISCARD_EXT`` to make it possible to disable
+them. This can be done by defining ``_LIBCPP_DISABLE_NODISCARD_EXT``.
+
+When should ``[[nodiscard]]`` be added to functions?
+====================================================
+
```
[the remark this answers is not in the document]
Comment (philnik777) [reply libcxx/docs/DesignDocs/NodiscardExtensions.rst:12]: This would fall under "where discarding the value is most likely a misuse of the function". IMO we should diagnose these kinds of cases. In your example `any_of`, or if you want to keep identical semantics `for_each`, would be the correct algorithm to use. I think we want to drive people towards writing better code, and this helps. Also, although it's not not _super_ relevant in this case, the side effect some program is relying on may be the result of some library-internal decision that could change as the library evolves. Avoiding people writing code like that is both in their interest and ours.

[the remark this answers is not in the document]
Comment (jwakely) [reply libcxx/docs/DesignDocs/NodiscardExtensions.rst:12]: IMHO if you're using the function like that then I think being forced to explicitly cast the result to void makes it clearer for readers that you're using it in an unusual way.

[the remark this answers is not in the document]
Comment (mordante) [reply libcxx/docs/DesignDocs/NodiscardExtensions.rst:12]: I don't feel it's poor style per se. In this case you could use `std::expected`. But I've seen similar creative usage of `std` algorithms in code bases.
IMO the cast to void is a hint to the reader something odd is going on and the algorithm is used in a different way. So I'm still happy to make these `[[nodiscard]]`.

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 3]
```diff
@@ -0,0 +1,34 @@
+======================================
```
Comment (ldionne) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:1]: I would call this file `NodiscardPolicy` instead.

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+``[[nodiscard]]`` extensions in libc++
```
Comment (ldionne) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:2]:
```suggestion
Guidelines for applying ``[[nodiscard]]`` in libc++
```

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+======================================
+
+Libc++ adds ``[[nodiscard]]`` to functions in a lot more places than the
```
Comment (ldionne) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:5]: Generally speaking this wording needs to be rebased on top of the Tokyo discussion + policy paper.

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+standard written as ``_LIBCPP_NODISCARD_EXT`` to make it possible to disable
+them. This can be done by defining ``_LIBCPP_DISABLE_NODISCARD_EXT``.
+
+When should ``[[nodiscard]]`` be added to functions?
+====================================================
+
+``[[nodiscard]]`` should be applied to functions
+
+- where it is most likely a correctness issue when discarding the return value.
+  For example a locking constructor in ``unique_lock``.
```
Comment (ldionne) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:16]:
```suggestion
- where discarding the return value is most likely a correctness issue.
  For example a locking constructor in ``unique_lock``.
```

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+- where most likely something similar was meant if the return value is
+  discarded. For example ``vector::empty()``, which probably should have
+  been ``clear()``.
```
Comment (ldionne) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:19]:
```suggestion
- where discarding the return value likely points to the user wanting to do something different. For example ``vector::empty()``, which probably should have
  been ``vector::clear()``.
```

File: libcxx/docs/DesignDocs/NodiscardExtensions.rst [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  to find.
+
+- which return a constant. For example ``numeric_limits::min()``.
+- which only observe a value. For example ``string::size()``.
+
+  Code that discards values from these kinds of functions is dead code. It can
+  either be removed, or the programmer meant to do something different.
+
+- where discarding the value is most likely a misuse of the function. For
+  example ``find``.
```
Comment (ldionne) [inline libcxx/docs/DesignDocs/NodiscardExtensions.rst:31]: Not attached to this line: this document should touch on the testing expectations (and briefly how to do it without getting into any details). I'd expect something like

> Applications of `[[nodiscard]]` are code like any other code, so we aim to test them. This can be done with a `.verify.cpp` test and many examples are available (look for tests named `xxxxx.nodiscard.verify.cpp`)

**Review by ldionne — APPROVED** [review]
LGTM with some comments. I'd also expect another PR that cleans up the current state of `_LIBCPP_NODISCARD_EXT`.

**Comment by philnik777** [comment]
For reference, https://github.com/microsoft/STL/issues/206 lists when MSVC applies `[[nodiscard]]`, which is very similar to the list proposed here.

**Comment by philnik777** [comment]
If anybody wants to add any more places we should add `[[nodiscard]]` feel free to open a new PR. I didn't see any feedback against the places proposed here, so I'm landing this now.
